src/os_posix/os_union_fs.c

```diff
@@ -536,168 +540,139 @@ __union_fs_file_lock(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, bool l
-    if ((alloc = union_fh->destination.allocation_list) == NULL)
-        return (NONE);
+    /* FIXME - This 100% needs concurrency control. Locking is easy, but a CAS might be straight
+     * forward?
+     */
+    hole = union_fh->destination.hole_list;
+    prev_hole = NULL;
+    while (hole != NULL) {
 
-    while (alloc != NULL) {
-        /* The read is in this allocation. */
-        if (offset >= alloc->off && rw_end <= EXTENT_END(alloc)) {
-            __wt_verbose_debug3(
-              session, WT_VERB_FILEOPS, "CAN SERVICE %s: Full match", union_fh->iface.name);
-            return (FULL);
+        if (write_end < hole->off) {
+            /* We won't find any more overlapping holes. Stop searching. */
+            break;
         }
-        alloc = alloc->next;
-    }
 
-    return (NONE);
-}
+        if (offset <= hole->off && write_end >= EXTENT_END(hole)) {
+            /* The write fully overlaps a hole. Delete it. */
+            __wt_verbose_debug3(
+              session, WT_VERB_FILEOPS, "Fully overlaps hole %ld-%ld", hole->off, EXTENT_END(hole));
+
+            tmp = hole;
+            if (prev_hole == NULL)
+                union_fh->destination.hole_list = hole->next;
+            else
+                prev_hole->next = hole->next;
+            hole = hole->next;
+            __wt_free(session, tmp);
+            continue;
 
-/*
- * __union_merge_with_next_extents --
- *     Merge the current extent with the following extent(s) if they overlap.
- */
-static int
-__union_merge_with_next_extents(WT_SESSION_IMPL *session, WT_UNION_ALLOC_LIST *extent)
-{
-    WT_UNION_ALLOC_LIST *next_extent;
-    wt_off_t new_len;
-
-    next_extent = extent->next;
-    WT_ASSERT_ALWAYS(session, next_extent != NULL, "Attempting to merge with NULL!");
-    WT_ASSERT_ALWAYS(session, EXTENT_END(extent) >= next_extent->off,
-      "Attempting to merge non-overlapping extents! extent_end = %ld, next->off = %ld",
-      EXTENT_END(extent), next_extent->off);
-
-    extent->next = next_extent->next;
-    new_len = EXTENT_END(next_extent) - extent->off; // TODO check for off by one
-    // FIXME - A lot of pointless typecasting. Consider dropping wt_off_t in favor of size_t.
-    extent->len = (size_t)WT_MAX((wt_off_t)extent->len, new_len);
-    __wt_free(session, next_extent);
-
-    // Run it again in case we also overlap with the N+1 extent
-    // TODO - loop instead of recursion.
-    if (extent->next != NULL && EXTENT_END(extent) >= extent->next->off) {
-        WT_RET(__union_merge_with_next_extents(session, extent));
+        } else if (offset > hole->off && write_end < EXTENT_END(hole)) {
+            /* The write is entirely within the hole. Split the hole in two. */
+
+            __wt_verbose_debug3(session, WT_VERB_FILEOPS, "Fully contained by hole %ld-%ld",
+              hole->off, EXTENT_END(hole));
+
+            /* First create the hole to the right of the write. */
+            WT_RET(__wt_calloc_one(session, &new));
```

Comment:
We should have a func for allocating extents in the event that we do not end up using WT_EXTs.

Reply:
Added in [eb93c7b](https://github.com/wiredtiger/wiredtiger/pull/11268/commits/eb93c7b3a81f1f85b40729e50f99f41a3944cc2b)